tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc

```diff
@@ -3059,5 +3103,166 @@ Status IrEmitterUnnested::EmitConstantGlobals() {
+
+  HloInstruction* slice_or_tuple = unnested_hlo->fused_expression_root();
+  auto slice_instructions = ([&]() -> absl::Span<HloInstruction* const> {
+    if (slice_or_tuple->opcode() == HloOpcode::kSlice) {
+      return absl::Span<HloInstruction* const>(&slice_or_tuple, 1);
+    }
+    DCHECK(slice_or_tuple->opcode() == HloOpcode::kTuple);
+    return slice_or_tuple->operands();
+  })();
+
+  // Emit input operand values of slices here.
+  std::vector<llvm::Value*> input_ir_values;
+  GpuElementalIrEmitter elem_emitter(hlo_module_config_, module_, &b_,
+                                     GetNestedComputer());
+  FusedIrEmitter fused_emitter(GetGeneratorForOperandIrArrays(unnested_hlo),
+                               &elem_emitter);
+  TF_CHECK_OK(unnested_hlo->fused_expression_root()->Accept(&fused_emitter));
+  for (size_t i = 0; i < slice_instructions.size(); ++i) {
+    const HloInstruction* slice = slice_instructions[i];
+    auto input_generator = fused_emitter.GetGenerator(slice->operand(0));
+    input_ir_values.push_back(input_generator(index).ValueOrDie());
+  }
+
+  // Emit for slice_instructions begins here.
+  llvm::BasicBlock* init_block = b_.GetInsertBlock();
+  llvm::BasicBlock* exit_block;
+  exit_block = init_block->splitBasicBlock(b_.GetInsertPoint(),
+                                           IrName(unnested_hlo, "merge"));
+
+  // First, generate blocks and guard_blocks for each slice. Will populate
+  // them below.
+  std::vector<llvm::BasicBlock*> emitted_guard_blocks;
+  std::vector<llvm::BasicBlock*> emitted_slice_blocks;
+  for (size_t i = 0; i < slice_instructions.size(); ++i) {
+    emitted_guard_blocks.emplace_back(llvm_ir::CreateBasicBlock(
+        exit_block, StrCat("slice_guard_id", i), &b_));
+    emitted_slice_blocks.emplace_back(
+        llvm_ir::CreateBasicBlock(exit_block, StrCat("slice_id", i), &b_));
+  }
+
+  // Jump from init_block to the first guard_block.
+  init_block->getTerminator()->eraseFromParent();
+  b_.SetInsertPoint(init_block);
+  Br(emitted_guard_blocks[0]);
+
+  // Populate the guard blocks.
+  for (size_t i = 0; i < slice_instructions.size(); ++i) {
+    b_.SetInsertPoint(emitted_guard_blocks[i]);
+    const HloInstruction* slice = slice_instructions[i];
+
+    // guarding_cond := index >= start && index < limit.
+    llvm::Value* guarding_cond = nullptr;
+    for (size_t dim = 0; dim < slice->slice_starts().size(); ++dim) {
+      DCHECK(slice->slice_strides(dim) == 1);
+      auto larger_or_equal_than_start =
+          ICmpSGE(index.multidim()[dim],
+                  index.GetConstantWithIndexType(slice->slice_starts(dim)));
+      auto smaller_than_limit =
+          ICmpSLT(index.multidim()[dim],
+                  index.GetConstantWithIndexType(slice->slice_limits(dim)));
+      auto within_range = And(larger_or_equal_than_start, smaller_than_limit);
+      guarding_cond = (guarding_cond == nullptr)
+                          ? within_range
```

Comment:
It seems cleaner to collect all guards in a vector first, and then use `b_.CreateAnd(conds)`

Reply:
Good suggestion.